dockerproxy/auth.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/superfly/flyctl/api"
+)
+
+func authRequest(next http.Handler) http.Handler {
+	if noAuth {
+		return next
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		appName, authToken, ok := r.BasicAuth()
+
+		if !ok || !authorizeRequestWithCache(appName, authToken) {
+			w.WriteHeader(http.StatusUnauthorized)
+			err := json.NewEncoder(w).Encode(map[string]string{
+				"message": "You are not authorized to use this builder",
+			})
+			if err != nil {
+				log.Warnln("error writing response", err)
+			}
+			return
+		}
+
+		next.ServeHTTP(w, r)
+	})
+}
+
+func authorizeRequestWithCache(appName, authToken string) bool {
+	if noAuth {
+		return true
+	}
+
+	if appName == "" || authToken == "" {
+		return false
+	}
+
+	cacheKey := appName + ":" + authToken
+	if val, ok := authCache.Get(cacheKey); ok {
+		if authorized, ok := val.(bool); ok {
+			log.Debugln("authorized from cache")
+			return authorized
+		}
+	}
+
+	authorized := authorizeRequest(appName, authToken)
+	authCache.Set(cacheKey, authorized, 0)
+	log.Debugln("authorized from api")
+	return authorized
+}
+
+// TODO: If we know that we're always going to use 6pn to access builders, we can probably just drop this auth since the network will take care to authorize access within the same org?
+func authorizeRequest(appName, authToken string) bool {
+	fly := api.NewClient(authToken, fmt.Sprintf("superfly/rchab/%s", gitSha), "0.0.0.0.0.0.1", log)
+
+	app, err := fly.GetAppCompact(context.TODO(), appName)
+	if app == nil || err != nil {
+		log.Warnf("Error fetching app %s: %v", appName, err)
+		return false
+	}
```

Comment:
You could add a `ctx` argument in this function and pass in `r.Context()` from the caller so you don't need to use `context.TODO()`.